Migration to remove erroneous tx state

brad-decker opened the pull request:
requires: #11103

brad-decker committed:
migration to remove orphaned cancel/retry txs

moymoy08 commented:
https://www.contributor-covenant.org/version/2/0/code_of_conduct.html

moymoy08 commented:
https://www.contributor-covenant.org/version/2/0/code_of_conduct.html

moymoy08 commented:
https://www.contributor-covenant.org/version/2/0/code_of_conduct.html

danjm commented on app/scripts/migrations/059.js:
```diff
@@ -0,0 +1,50 @@
+  const transactions = state?.TransactionController?.transactions;
+  if (
+    typeof transactions === 'object' &&
+    Array.isArray(transactions) === false &&
```
Why the need to verify that it is not an array? And is any update to state needed when it is an array?

danjm replied:
Also, lodash can help simplify this:

![Screenshot from 2021-05-17 11-44-09](https://user-images.githubusercontent.com/7499938/118503514-437e4400-b705-11eb-9fb9-21f113439832.png)

brad-decker replied:
If transactions are an array here it'll mean that the state is corrupt because in a previous migration we migrated the transaction state store to an id indexed object. So if its an array this point the user has bigger problems. I added these checks to make the code as verbose as possible. There are also corresponding tests to see what would happen in the case where it isn't in the proper state.

brad-decker replied:
added isPlainObject in latest

danjm commented on app/scripts/migrations/059.test.js:
```diff
@@ -0,0 +1,384 @@
+import { TRANSACTION_TYPES } from '../../../shared/constants/transaction';
+import migration59 from './059';
+
+const ERRONEOUS_TRANSACTION_STATE = {
```
Can we add a case to these tests that where there is a transaction with a 'cancel' type, but it shares a nonce with a non-cancel transaction, to verify that it won't be removed?

brad-decker replied:
there is one. line 192

```js
 it('should not drop cancel transactions with matching non cancel or retry in same network and nonce', async function () {
    const oldStorage = {
      meta: {},
      data: {
        TransactionController: {
          transactions: {
            ...ERRONEOUS_TRANSACTION_STATE,
            11: {
              ...ERRONEOUS_TRANSACTION_STATE['0'],
              id: 11,
              type: TRANSACTION_TYPES.SENT_ETHER,
            },
          },
        },
        foo: 'bar',
      },
    };

    const newStorage = await migration59.migrate(oldStorage);
    assert.deepEqual(newStorage.data, {
      TransactionController: {
        transactions: oldStorage.data.TransactionController.transactions,
      },
      foo: 'bar',
    });
  });
```

Gudahtt reviewed (commented):
This doesn't seem to account for pending transactions :thinking: 

I'm not entirely sure what we should do for those though. I was thinking that we could exclude them from the migration altogether, but what about the case where there's a pending "CANCEL" with no base transaction from which to infer the original intent?

brad-decker commented:
> I'm not entirely sure what we should do for those though. I was thinking that we could exclude them from the migration altogether, but what about the case where there's a pending "CANCEL" with no base transaction from which to infer the original intent?

if they have no base transaction to infer from they need to be removed because they are part of the erroneous state... the only alternative would be to change them to ether transaction which would potentially confuse the user. This does seem to be an edge case though because when creating the cancel transaction the original should be in pending status (and immune from excision), and the status of the original should only change if it is A) confirmed (in which case the cancel will be dropped) of B) dropped (when the cancel in confirmed)...

Gudahtt commented:
Ah, right good point. OK, in that case it seems fair to treat them the same as finished transactions.

I just thought of another potential complication: Unapproved transactions. They typically don't have a nonce yet, so they'll be in a nonce group with `undefined` in it. Which.... is probably fine? They can't possibly be cancel/retry so they won't get filtered. But it might be worth adding a test at least to confirm they don't blow anything up.

brad-decker committed:
wrench monkey: nonce-less unapproved tx

brad-decker commented:
I changed one transaction in the ERRONEOUS_TRANSACTION_STATE to be an unapproved, nonce-less tx. I actually just deleted the txParams object entirely which did result in an error so I added optional chaining on the txParams.nonce in the migration.

brad-decker commented:
any hints on the benchmark fail?

Gudahtt commented:
That's an intermittent failure specific to chromedriver. It happens sporadically on the chrome e2e tests as well, except we have a retry for those, so we don't see it bubble up as often.

Gudahtt reviewed (approved):
LGTM!
